File: backend/src/main/java/com/microfinance/code/model/SMELoan.java
package com.microfinance.code.model;

import com.microfinance.code.status.LoanStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@Table(name = "sme_loan")
public class SMELoan {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;

    @Column(name = "loan_id", nullable = false, length = 30) // Added nullable and length
    private String loanId;

    @Column(name = "loan_amount", nullable = false) // Added nullable
    private BigDecimal loanAmount;

    @Column(name = "interest_rate", nullable = false) // Added nullable
    private BigDecimal interestRate;

    @Column(name = "grace_period", nullable = true) // Added nullable
    private int gracePeriod;

    @Column(name = "loan_purpose", nullable = false, length = 200) // Added nullable and length
    private String loanPurpose;

    @Column(name = "registered_date", nullable = false) // Added nullable
    private LocalDateTime registeredDate;

    @Column(name = "approved_date") // approveDate can be null initially
    private LocalDateTime approvedDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private LoanStatus status;

    @Column(name = "document_fee", nullable = false) // Added nullable
    private BigDecimal documentFee;

    @Column(name = "service_charge", nullable = false) // Added nullable
    private BigDecimal serviceCharge;

    @Column(name = "expired_date", nullable = true) // Added nullable
    private LocalDateTime expiredDate;

    @Column(name = "duration", nullable = false) // Added nullable
    private int duration;

    @Column(name = "principal", nullable = false) // Added nullable
    private BigDecimal principal;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User entryUser;

    @ManyToOne
    @JoinColumn(name = "approvedUser_id", nullable = false)
    private User approvedUser;

    @ManyToOne
    @JoinColumn(name = "current_account_id", nullable = false)
    private CurrentAccount currentAccount;

    @PrePersist
    public void prePersist() {
        if (this.status == null) { // Only set if not already set
            this.status = LoanStatus.PENDING;
        }
        if(registeredDate == null){
            LocalDateTime now = LocalDateTime.now();
            registeredDate = now ;
        }
        if(principal == null){
            principal = loanAmount;
        }
    }
}

File: frontend/MicroFinance_APP/src/app/model/SmeLoan.ts
import { CollateralDTO } from "./CollateralDTO";

export interface Smeloan {
  id?: number;
  loanId?: string;              // Corresponds to loanId in Java DTO
  loanAmount?: number;          // Corresponds to loanAmount in Java DTO
  interestRate?: number;        // Corresponds to interestRate in Java DTO
  gracePeriod?: number;         // Corresponds to gracePeriod in Java DTO (you might not have it in the TypeScript DTO)
  loanPurpose?: string;         // Corresponds to loanPurpose in Java DTO
  registeredDate?: string;      // Corresponds to registeredDate (String) in Java DTO
  approvedDate: string;         // Corresponds to approvedDate (String) in Java DTO
  status: string;               // Corresponds to status in Java DTO
  documentFee?: number;         // Corresponds to documentFee in Java DTO (if not needed, leave it out)
  serviceCharge?: number;       // Corresponds to serviceCharge in Java DTO (if not needed, leave it out)
  collateralIds?: number[];     // Corresponds to collateralIds (List<Integer>) in Java DTO
  expiredDate?: string;         // Corresponds to expiredDate (LocalDate) in Java DTO, use String here
  duration?: number;            // Corresponds to duration in Java DTO
  principal?: number;           // Corresponds to principal in Java DTO
  entryUserId?: number;         // Corresponds to entryUserId in Java DTO
  entryUserGenerateId?: string; // Corresponds to entryUserGenerateId in Java DTO (you might not have this field)
  entryUserName?: string;       // Corresponds to entryUserName in Java DTO
  approvedUserId?: number;      // Corresponds to approvedUserId in Java DTO
  approvedUserName?: string;    // Corresponds to approvedUserName in Java DTO
  currentAccountId?: number;    // Corresponds to currentAccountId in Java DTO
  currentAccountaccId?: string; // Corresponds to currentAccountaccId in Java DTO
  borrowerName?: string;        // Corresponds to borrowerName in Java DTO
  cifId: number;
  cifIdNumber: string;
  usedCollaterals: CollateralDTO[];
  loanStatus: string;
}
